// nestjs-seeder/dist/seeder/seeder.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var SeederModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeederModule = void 0;
const common_1 = require("@nestjs/common");
const seeder_service_1 = require("./seeder.service");
let SeederModule = SeederModule_1 = class SeederModule {
    static register(options) {
        return {
            module: SeederModule_1,
            imports: options.imports || [],
            providers: [
                ...(options.providers || []),
                ...options.seeders,
                {
                    provide: seeder_service_1.SeederService,
                    useFactory: (...seeders) => {
                        return new seeder_service_1.SeederService(seeders);
                    },
                    inject: options.seeders,
                },
            ],
        };
    }
};
SeederModule = SeederModule_1 = __decorate([
    (0, common_1.Module)({})
], SeederModule);
exports.SeederModule = SeederModule;
